src/parser/core/modules/BuffWindow.tsx

```diff
@@ -0,0 +1,329 @@
+
+class BuffWindowState {
+	start: number
+	end?: number
+	rotation: CastEvent[] = []
+
+	constructor(start: number) {
+		this.start = start
+	}
+
+	get gcds(): number {
+		return this.rotation
+			.map(e => getDataBy(ACTIONS, 'id', e.ability.guid) as TODO)
+			.filter(a => a && a.onGcd)
+			.length
+	}
+
+	getActionCountByIds(actionsById: number[]): number {
+		return this.rotation
+			.filter(e => actionsById.includes(e.ability.guid))
+			.length
+	}
+}
+
+interface SeverityTiers {
+	[key: number]: number
+}
+
+interface BuffWindowExpectedGCDs {
+	expectedPerWindow: number
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+interface BuffWindowRequiredGCDs {
+	iconAction: ActionData
+	actions: ActionData[]
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+interface BuffWindowTrackedCooldown {
+	action: ActionData
+	expectedPerWindow: number,
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+export default class BuffWindowModule extends Module {
+	static handle: string = 'buffwindow'
+	static title: MessageDescriptor = t('core.buffwindow.title')`Buff Window`
+
+	// Implementing modules should pass the ACTION and STATUS object respectively that represents the buff window
+	protected buffAction?: ActionData
+	protected buffStatus?: StatusData
+
+	// From a practical perspective, expectedGCDs should be considered required for implementing modules.
+	// Other properties can be added as desired.  trackedBadCooldowns will show suggestions if used too many times, but will not show in the rotation tables
+	protected expectedGCDs?: BuffWindowExpectedGCDs
+	protected requiredGCDs?: BuffWindowRequiredGCDs
+	protected trackedCooldowns?: BuffWindowTrackedCooldown[]
+	protected trackedBadCooldowns?: BuffWindowTrackedCooldown[]
+
+	@dependency private suggestions!: Suggestions
+	@dependency private timeline!: Timeline
+	@dependency private globalCooldown!: GlobalCooldown
+
+	private buffWindows: BuffWindowState[] = []
+
+	private get activeBuffWindow(): BuffWindowState | undefined {
+		const lastBuffWindow = _.last(this.buffWindows)
+		if ( lastBuffWindow && lastBuffWindow.end == null ) {
+			return lastBuffWindow
+		}
+		return undefined
+	}
+
+	protected init() {
+		this.addHook('cast', {by: 'player'}, this.onCast)
+		this.addHook('applybuff', {by: 'player'}, this.onApplyBuff)
+		this.addHook('removebuff', {by: 'player'}, this.onRemoveBuff)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onCast(event: CastEvent) {
+		const action: ActionData | undefined = getDataBy(ACTIONS, 'id', event.ability.guid)
+
+		if (!action || action.autoAttack) {
+			// Disregard auto attacks for tracking rotations / events during buff windows
+			return
+		}
+
+		if (this.activeBuffWindow) {
+			this.activeBuffWindow.rotation.push(event)
+		}
+	}
+
+	private onApplyBuff(event: BuffEvent) {
+		if (!this.buffStatus || event.ability.guid !== this.buffStatus.id) {
```

Comment:
Unless it's expected that the buff status will be changing on a regular basis, this id check should be in the `init` as a filter param.

Reply:
_Especially_ if the class is an abstract class to enforce buffAction & buffStatus properties, I can't put the ID check in the init: https://github.com/Microsoft/TypeScript/issues/15247

I didn't have much luck with putting the ID check in the init even after I moved away from an abstract class/props, so I left it this way.

Reply:
Bleh i keep forgetting that